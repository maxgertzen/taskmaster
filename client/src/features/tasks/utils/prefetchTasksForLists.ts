import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
