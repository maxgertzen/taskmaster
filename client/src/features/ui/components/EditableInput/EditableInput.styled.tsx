import styled from '@emotion/styled';

export const StyledEditableInput = styled.input(({ theme }) => ({
  display: 'flex',
  gap: theme.spacing(1),
}));
