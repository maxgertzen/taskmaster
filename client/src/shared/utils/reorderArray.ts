export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
