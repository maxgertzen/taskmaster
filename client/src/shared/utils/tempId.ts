export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
