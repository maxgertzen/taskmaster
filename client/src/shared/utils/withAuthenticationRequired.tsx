import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
