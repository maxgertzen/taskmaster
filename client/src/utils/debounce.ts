export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
