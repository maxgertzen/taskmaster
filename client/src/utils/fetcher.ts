export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
