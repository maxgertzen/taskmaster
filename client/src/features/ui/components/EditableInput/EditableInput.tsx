import { FC, useState } from 'react';

import { StyledEditableInput } from './EditableInput.styled';

interface EditableInputProps {
  initialText?: string;
  placeholder: string;
  onSubmit: (text: string) => void;
  onCancel: () => void;
}

export const EditableInput: FC<EditableInputProps> = ({
  initialText = '',
  placeholder,
  onSubmit,
  onCancel,
}) => {
  const [text, setText] = useState(initialText);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setText(event.target.value);
  };

  const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      handleSubmit();
    } else if (event.key === 'Escape') {
      onCancel();
    }
  };

  const handleBlur = () => {
    handleSubmit();
  };

  const handleSubmit = () => {
    if (text.trim()) {
      onSubmit(text.trim());
    } else {
      onCancel();
    }
    setText('');
  };

  return (
    <StyledEditableInput
      value={text}
      onChange={handleChange}
      onKeyDown={handleKeyPress}
      onBlur={handleBlur}
      placeholder={placeholder}
      autoFocus
    />
  );
};
