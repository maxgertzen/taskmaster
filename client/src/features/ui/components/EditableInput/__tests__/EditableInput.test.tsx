import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import { render, screen } from '@/tests/utils';

import { EditableInput } from '../EditableInput';

describe('EditableInput', () => {
  const mockSubmit = vi.fn();
  const mockCancel = vi.fn();

  beforeEach(() => {
    mockSubmit.mockClear();
    mockCancel.mockClear();
  });

  it('renders correctly in light theme', () => {
    const { container } = render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />,
      { theme: 'light' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly in dark theme', () => {
    const { container } = render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />,
      { theme: 'dark' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with initial text when provided', () => {
    render(
      <EditableInput
        initialText='Initial value'
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByDisplayValue('Initial value');
    expect(input).toBeInTheDocument();
  });

  it('updates text on change', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');
    act(() => {
      fireEvent.change(input, { target: { value: 'New text' } });
    });

    expect(input).toHaveValue('New text');
  });

  it('calls onSubmit with trimmed text when Enter is pressed', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');

    act(() => {
      fireEvent.change(input, { target: { value: '  Submit this  ' } });
      fireEvent.keyDown(input, { key: 'Enter' });
    });

    expect(mockSubmit).toHaveBeenCalledWith('Submit this');
    expect(input).toHaveValue('');
  });

  it('calls onCancel when Escape is pressed', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');
    act(() => {
      fireEvent.change(input, { target: { value: 'Cancel this' } });
      fireEvent.keyDown(input, { key: 'Escape' });
    });

    expect(mockCancel).toHaveBeenCalled();
    expect(mockSubmit).not.toHaveBeenCalled();
  });

  it('calls onSubmit with trimmed text on blur when text exists', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');

    act(() => {
      fireEvent.change(input, { target: { value: '  Blur text  ' } });
      input.blur();
    });

    expect(mockSubmit).toHaveBeenCalledWith('Blur text');
    expect(input).toHaveValue('');
  });

  it('calls onCancel on blur when text is empty or only whitespace', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');
    act(() => {
      fireEvent.change(input, { target: { value: '   ' } });
      input.blur();
    });

    expect(mockCancel).toHaveBeenCalled();
    expect(mockSubmit).not.toHaveBeenCalled();
    expect(input).toHaveValue('');
  });

  it('has autofocus by default', () => {
    render(
      <EditableInput
        placeholder='Type something'
        onSubmit={mockSubmit}
        onCancel={mockCancel}
      />
    );

    const input = screen.getByPlaceholderText('Type something');
    expect(input).toHaveFocus();
  });
});
