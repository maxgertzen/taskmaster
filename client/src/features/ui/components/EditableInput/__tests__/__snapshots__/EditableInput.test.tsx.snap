// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableInput > renders correctly in dark theme 1`] = `
<div>
  <input
    class="css-1i27l4i"
    data-testid="editable-input"
    placeholder="Type something"
    value=""
  />
</div>
`;

exports[`EditableInput > renders correctly in light theme 1`] = `
<div>
  <input
    class="css-1i27l4i"
    data-testid="editable-input"
    placeholder="Type something"
    value=""
  />
</div>
`;
